// @aws-amplify/ui-react-native/lib/InAppMessaging/components/ModalMessage/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLandscapeStyles = exports.getPortraitStyles = void 0;
const react_native_1 = require("react-native");
const constants_1 = require("../../constants");
const commonStyles = {
    body: {
        fontSize: constants_1.FONT_SIZE_BASE,
        fontWeight: constants_1.FONT_WEIGHT_BASE,
        lineHeight: constants_1.LINE_HEIGHT_BASE,
    },
    buttonContainer: {
        backgroundColor: constants_1.COLOR_LIGHT_GREY,
        borderRadius: constants_1.BORDER_RADIUS_BASE,
        flex: 1,
        margin: constants_1.SPACING_MEDIUM,
        padding: constants_1.SPACING_LARGE,
    },
    buttonsContainer: {
        flexDirection: 'row',
        justifyContent: 'center',
    },
    container: {
        backgroundColor: constants_1.COLOR_WHITE,
        elevation: constants_1.MESSAGE_ELEVATION,
        margin: constants_1.SPACING_EXTRA_LARGE,
        shadowColor: constants_1.COLOR_BLACK,
        shadowOffset: {
            width: constants_1.MESSAGE_SHADOW_WIDTH,
            height: constants_1.MESSAGE_SHADOW_HEIGHT,
        },
        shadowOpacity: constants_1.MESSAGE_SHADOW_OPACITY,
        shadowRadius: constants_1.MESSAGE_SHADOW_RADIUS,
    },
    buttonText: {
        fontSize: constants_1.FONT_SIZE_BASE,
        fontWeight: constants_1.FONT_WEIGHT_BASE,
        lineHeight: constants_1.LINE_HEIGHT_BASE,
        textAlign: 'center',
    },
    header: {
        fontSize: constants_1.FONT_SIZE_LARGE,
        fontWeight: constants_1.FONT_WEIGHT_BOLD,
        lineHeight: constants_1.LINE_HEIGHT_LARGE,
    },
    iconButton: {
        alignSelf: 'flex-start',
        marginLeft: 'auto',
    },
    imageContainer: {
        alignItems: 'center',
        margin: constants_1.SPACING_LARGE,
    },
    wrapper: {
        flex: 1,
        justifyContent: 'center',
    },
};
const getPortraitStyles = (imageDimensions) => react_native_1.StyleSheet.create({
    ...commonStyles,
    buttonsContainer: {
        ...commonStyles.buttonsContainer,
        marginTop: 'auto',
        paddingHorizontal: constants_1.SPACING_SMALL,
        paddingBottom: constants_1.SPACING_SMALL,
    },
    container: {
        ...commonStyles.container,
        minHeight: '40%',
    },
    contentContainer: { padding: constants_1.SPACING_LARGE },
    image: { ...imageDimensions },
    textContainer: {
        marginTop: constants_1.SPACING_LARGE,
    },
});
exports.getPortraitStyles = getPortraitStyles;
const getLandscapeStyles = (imageDimensions) => react_native_1.StyleSheet.create({
    ...commonStyles,
    container: {
        ...commonStyles.container,
        flex: 1,
        padding: constants_1.SPACING_EXTRA_LARGE,
    },
    contentContainer: {
        flex: 1,
        flexDirection: 'row',
    },
    iconButton: {
        ...commonStyles.iconButton,
        marginRight: constants_1.SPACING_MEDIUM,
    },
    image: { ...imageDimensions },
    imageContainer: {
        ...commonStyles.imageContainer,
        justifyContent: 'center',
    },
    textContainer: {
        flex: 1,
        justifyContent: 'center',
        paddingHorizontal: constants_1.SPACING_MEDIUM,
    },
});
exports.getLandscapeStyles = getLandscapeStyles;
