// @aws-amplify/ui-react-native/lib/InAppMessaging/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IN_APP_MESSAGING_TEST_ID = exports.BUTTON_PRESSED_OPACITY = exports.MESSAGE_ELEVATION = exports.MESSAGE_SHADOW_RADIUS = exports.MESSAGE_SHADOW_OPACITY = exports.MESSAGE_SHADOW_WIDTH = exports.MESSAGE_SHADOW_HEIGHT = exports.ICON_BUTTON_HIT_SLOP = exports.ICON_BUTTON_SIZE = exports.FONT_WEIGHT_BOLD = exports.FONT_WEIGHT_BASE = exports.LINE_HEIGHT_LARGE = exports.LINE_HEIGHT_BASE = exports.FONT_SIZE_LARGE = exports.FONT_SIZE_BASE = exports.BORDER_RADIUS_BASE = exports.SPACING_EXTRA_LARGE = exports.SPACING_LARGE = exports.SPACING_MEDIUM = exports.SPACING_SMALL = exports.COLOR_WHITE = exports.COLOR_LIGHT_GREY_2 = exports.COLOR_LIGHT_GREY = exports.COLOR_GREY = exports.COLOR_BLACK = void 0;
const utils_1 = require("../utils");
// color
exports.COLOR_BLACK = '#000';
exports.COLOR_GREY = '#a1a1a1';
exports.COLOR_LIGHT_GREY = '#e8e8e8';
exports.COLOR_LIGHT_GREY_2 = '#d8d8d8';
exports.COLOR_WHITE = '#fff';
// spacing
exports.SPACING_SMALL = 4;
exports.SPACING_MEDIUM = 8;
exports.SPACING_LARGE = 12;
exports.SPACING_EXTRA_LARGE = 16;
// border radius
exports.BORDER_RADIUS_BASE = 4;
// font
exports.FONT_SIZE_BASE = 16;
exports.FONT_SIZE_LARGE = 18;
exports.LINE_HEIGHT_BASE = (0, utils_1.getLineHeight)(exports.FONT_SIZE_BASE);
exports.LINE_HEIGHT_LARGE = (0, utils_1.getLineHeight)(exports.FONT_SIZE_LARGE);
exports.FONT_WEIGHT_BASE = '400';
exports.FONT_WEIGHT_BOLD = '600';
// icon
exports.ICON_BUTTON_SIZE = 20;
exports.ICON_BUTTON_HIT_SLOP = 10;
// shadow properties
// iOS shadow values
exports.MESSAGE_SHADOW_HEIGHT = 2;
exports.MESSAGE_SHADOW_WIDTH = 2;
exports.MESSAGE_SHADOW_OPACITY = 0.1;
exports.MESSAGE_SHADOW_RADIUS = 2;
// android shadow values
exports.MESSAGE_ELEVATION = 3;
// Message UI Buttons
// default value applied in React Native TouchableOpacity
exports.BUTTON_PRESSED_OPACITY = 0.8;
// Message Component Test IDs
exports.IN_APP_MESSAGING_TEST_ID = {
    BODY: 'in-app-messaging--body',
    CLOSE_BUTTON: 'in-app-messaging--close-button',
    HEADER: 'in-app-messaging--header',
    IMAGE: 'in-app-messaging--image',
    PRIMARY_BUTTON: 'in-app-messaging--primary-button',
    SECONDARY_BUTTON: 'in-app-messaging--secondary-button',
    CAROUSEL: 'in-app-messaging--carousel',
};
