// @aws-amplify/ui-react-native/lib/primitives/Heading/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getThemedStyles = void 0;
const react_native_1 = require("react-native");
const utils_1 = require("../../utils");
const getThemedStyles = (theme) => {
    const { components, tokens: { colors, fontSizes, fontWeights }, } = theme;
    return react_native_1.StyleSheet.create({
        text: {
            color: colors.font.primary,
            ...components?.heading?.text,
        },
        1: {
            fontSize: fontSizes.xxxl,
            fontWeight: fontWeights.light,
            lineHeight: (0, utils_1.getLineHeight)(fontSizes.xxxl),
            ...components?.heading?.[1],
        },
        2: {
            fontSize: fontSizes.xxl,
            fontWeight: fontWeights.normal,
            lineHeight: (0, utils_1.getLineHeight)(fontSizes.xxl),
            ...components?.heading?.[2],
        },
        3: {
            fontSize: fontSizes.xl,
            fontWeight: fontWeights.medium,
            lineHeight: (0, utils_1.getLineHeight)(fontSizes.xl),
            ...components?.heading?.[3],
        },
        4: {
            fontSize: fontSizes.large,
            fontWeight: fontWeights.medium,
            lineHeight: (0, utils_1.getLineHeight)(fontSizes.large),
            ...components?.heading?.[4],
        },
        5: {
            fontSize: fontSizes.medium,
            fontWeight: fontWeights.semibold,
            lineHeight: (0, utils_1.getLineHeight)(fontSizes.medium),
            ...components?.heading?.[5],
        },
        6: {
            fontSize: fontSizes.small,
            fontWeight: fontWeights.bold,
            lineHeight: (0, utils_1.getLineHeight)(fontSizes.small),
            ...components?.heading?.[6],
        },
    });
};
exports.getThemedStyles = getThemedStyles;
