// @aws-amplify/ui-react-native/lib/primitives/ErrorMessage/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getThemedStyles = void 0;
const react_native_1 = require("react-native");
const getThemedStyles = (theme) => {
    const { tokens: { colors, fontSizes, space }, components, } = theme;
    return react_native_1.StyleSheet.create({
        container: {
            alignItems: 'center',
            backgroundColor: colors.background.error,
            flexDirection: 'row',
            paddingHorizontal: space.xs,
            paddingVertical: space.small,
            ...components?.errorMessage?.container,
        },
        icon: {
            margin: space.xs,
            tintColor: colors.font.error,
            ...components?.errorMessage?.icon,
        },
        label: {
            color: colors.font.error,
            flex: 1,
            fontSize: fontSizes.medium,
            paddingHorizontal: space.xs,
            ...components?.errorMessage?.label,
        },
    });
};
exports.getThemedStyles = getThemedStyles;
