// @aws-amplify/ui-react-native/lib/Authenticator/common/DefaultContainer/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInnerContainerStyles = exports.getThemedStyles = void 0;
const react_native_1 = require("react-native");
const getThemedStyles = ({ tokens: { colors } }, insetPadding) => {
    return react_native_1.StyleSheet.create({
        keyboardAvoidingView: {
            alignContent: 'center',
            backgroundColor: colors.background.primary,
            justifyContent: 'center',
            flex: 1,
            // TDOO: add theme override
        },
        scrollViewContentContainer: {
            ...insetPadding,
            alignContent: 'center',
            justifyContent: 'center',
            flexGrow: 1,
            // TDOO: add theme override
        },
        scrollView: {}, // TDOO: add theme override
    });
};
exports.getThemedStyles = getThemedStyles;
const getInnerContainerStyles = ({ tokens: { space }, }) => react_native_1.StyleSheet.create({
    container: {
        marginHorizontal: space.medium,
        // TDOO: add theme override
    },
});
exports.getInnerContainerStyles = getInnerContainerStyles;
