// @aws-amplify/ui-react-native/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTheme = exports.defaultTheme = exports.defaultDarkModeOverride = exports.ThemeProvider = exports.withAuthenticator = exports.useAuthenticator = exports.Authenticator = void 0;
const tslib_1 = require("tslib");
var Authenticator_1 = require("./Authenticator");
Object.defineProperty(exports, "Authenticator", { enumerable: true, get: function () { return Authenticator_1.Authenticator; } });
Object.defineProperty(exports, "useAuthenticator", { enumerable: true, get: function () { return Authenticator_1.useAuthenticator; } });
Object.defineProperty(exports, "withAuthenticator", { enumerable: true, get: function () { return Authenticator_1.withAuthenticator; } });
tslib_1.__exportStar(require("./InAppMessaging"), exports);
var theme_1 = require("./theme");
Object.defineProperty(exports, "ThemeProvider", { enumerable: true, get: function () { return theme_1.ThemeProvider; } });
Object.defineProperty(exports, "defaultDarkModeOverride", { enumerable: true, get: function () { return theme_1.defaultDarkModeOverride; } });
Object.defineProperty(exports, "defaultTheme", { enumerable: true, get: function () { return theme_1.defaultTheme; } });
Object.defineProperty(exports, "useTheme", { enumerable: true, get: function () { return theme_1.useTheme; } });
