// @aws-amplify/ui-react-native/lib/InAppMessaging/hooks/useMessageImage/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODAL_IMAGE_SCREEN_SIZE = exports.FULL_SCREEN_IMAGE_SCREEN_SIZE = exports.CAROUSEL_IMAGE_SCREEN_SIZE = exports.BANNER_IMAGE_SCREEN_SIZE = exports.MODAL_IMAGE_SCREEN_MULTIPLIER = exports.FULL_SCREEN_IMAGE_SCREEN_MULTIPLIER = exports.CAROUSEL_IMAGE_SCREEN_MULTIPLIER = exports.BANNER_IMAGE_SCREEN_MULTIPLIER = void 0;
const react_native_1 = require("react-native");
// as images are not expected to be responsive to orientation changes get screen dimensions at app start
const SCREEN_DIMENSIONS = react_native_1.Dimensions.get('screen');
// compare screen width and height and assign the lesser of the two as the base screen dimension
const BASE_SCREEN_DIMENSION = SCREEN_DIMENSIONS.width < SCREEN_DIMENSIONS.height
    ? SCREEN_DIMENSIONS.width
    : SCREEN_DIMENSIONS.height;
// base size that message images should fill
// - all banner message images should fill 20 percent of the base screen dimension
// - all other components should fill 60 percent of the base screen dimension
exports.BANNER_IMAGE_SCREEN_MULTIPLIER = 0.2;
exports.CAROUSEL_IMAGE_SCREEN_MULTIPLIER = 0.6;
exports.FULL_SCREEN_IMAGE_SCREEN_MULTIPLIER = 0.6;
exports.MODAL_IMAGE_SCREEN_MULTIPLIER = 0.6;
exports.BANNER_IMAGE_SCREEN_SIZE = exports.BANNER_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;
exports.CAROUSEL_IMAGE_SCREEN_SIZE = exports.CAROUSEL_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;
exports.FULL_SCREEN_IMAGE_SCREEN_SIZE = exports.FULL_SCREEN_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;
exports.MODAL_IMAGE_SCREEN_SIZE = exports.MODAL_IMAGE_SCREEN_MULTIPLIER * BASE_SCREEN_DIMENSION;
