// @aws-amplify/ui-react-native/lib/primitives/Label/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getThemedStyles = void 0;
const react_native_1 = require("react-native");
const utils_1 = require("../../utils");
const getThemedStyles = (theme) => {
    const { components, tokens: { colors, fontSizes, fontWeights }, } = theme;
    return react_native_1.StyleSheet.create({
        text: {
            fontSize: fontSizes.medium,
            fontWeight: fontWeights.normal,
            lineHeight: (0, utils_1.getLineHeight)(fontSizes.medium),
            ...components?.label?.text,
        },
        primary: {
            color: colors.font.primary,
            ...components?.label?.primary,
        },
        secondary: {
            color: colors.font.secondary,
            ...components?.label?.secondary,
        },
        tertiary: {
            color: colors.font.tertiary,
            ...components?.label?.tertiary,
        },
        error: {
            color: colors.font.error,
            ...components?.label?.error,
        },
        warning: {
            color: colors.font.warning,
            ...components?.label?.warning,
        },
        success: {
            color: colors.font.success,
            ...components?.label?.success,
        },
        info: {
            color: colors.font.info,
            ...components?.label?.info,
        },
    });
};
exports.getThemedStyles = getThemedStyles;
