// @aws-amplify/ui-react-native/lib/Authenticator/common/DefaultContent/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultStyle = void 0;
const react_native_1 = require("react-native");
const getDefaultStyle = ({ tokens: { colors, fontSizes, space }, }) => react_native_1.StyleSheet.create({
    body: {
        fontSize: fontSizes.medium,
        paddingHorizontal: space.small,
    },
    buttonPrimary: {
        margin: space.small,
    },
    buttonPrimaryLabel: {},
    buttonSecondary: {
        marginHorizontal: space.medium,
    },
    buttonSecondaryLabel: {},
    errorMessage: {
        marginHorizontal: space.small,
    },
    errorMessageIcon: {},
    errorMessageLabel: {},
    fieldContainer: {
        paddingHorizontal: space.small,
    },
    fieldErrorsContainer: {
        paddingHorizontal: space.medium,
        paddingVertical: space.xs,
    },
    fieldError: {
        color: colors.font.error,
        paddingVertical: 1,
    },
    fieldLabel: {},
    fieldStyle: {},
    footer: {},
    formFields: {
        paddingBottom: space.xs,
    },
    header: {
        marginVertical: space.small,
        paddingHorizontal: space.small,
    },
    link: {
        marginVertical: space.xs,
        minWidth: '50%',
    },
    linksContainer: {
        flexDirection: 'row',
        flexWrap: 'wrap',
        justifyContent: 'center',
    },
});
exports.getDefaultStyle = getDefaultStyle;
