// @aws-amplify/react-native/dist/cjs/moduleLoaders/index.js
'use strict';

// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadAppState = exports.loadBase64 = exports.loadGetRandomValues = exports.loadUrlPolyfill = exports.loadBuffer = exports.loadNetInfo = exports.loadAsyncStorage = exports.loadAmplifyWebBrowser = exports.loadAmplifyPushNotification = void 0;
var loadAmplifyPushNotification_1 = require("./loadAmplifyPushNotification");
Object.defineProperty(exports, "loadAmplifyPushNotification", { enumerable: true, get: function () { return loadAmplifyPushNotification_1.loadAmplifyPushNotification; } });
var loadAmplifyWebBrowser_1 = require("./loadAmplifyWebBrowser");
Object.defineProperty(exports, "loadAmplifyWebBrowser", { enumerable: true, get: function () { return loadAmplifyWebBrowser_1.loadAmplifyWebBrowser; } });
var loadAsyncStorage_1 = require("./loadAsyncStorage");
Object.defineProperty(exports, "loadAsyncStorage", { enumerable: true, get: function () { return loadAsyncStorage_1.loadAsyncStorage; } });
var loadNetInfo_1 = require("./loadNetInfo");
Object.defineProperty(exports, "loadNetInfo", { enumerable: true, get: function () { return loadNetInfo_1.loadNetInfo; } });
var loadBuffer_1 = require("./loadBuffer");
Object.defineProperty(exports, "loadBuffer", { enumerable: true, get: function () { return loadBuffer_1.loadBuffer; } });
var loadUrlPolyfill_1 = require("./loadUrlPolyfill");
Object.defineProperty(exports, "loadUrlPolyfill", { enumerable: true, get: function () { return loadUrlPolyfill_1.loadUrlPolyfill; } });
var loadGetRandomValues_1 = require("./loadGetRandomValues");
Object.defineProperty(exports, "loadGetRandomValues", { enumerable: true, get: function () { return loadGetRandomValues_1.loadGetRandomValues; } });
var loadBase64_1 = require("./loadBase64");
Object.defineProperty(exports, "loadBase64", { enumerable: true, get: function () { return loadBase64_1.loadBase64; } });
var loadAppState_1 = require("./loadAppState");
Object.defineProperty(exports, "loadAppState", { enumerable: true, get: function () { return loadAppState_1.loadAppState; } });
