// @aws-amplify/ui-react-native/lib/primitives/Carousel/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VIEWABILITY_CONFIG = exports.DEFAULT_CAROUSEL_INDICATOR_INACTIVE_STYLE = exports.DEFAULT_CAROUSEL_INDICATOR_ACTIVE_STYLE = exports.DEFAULT_CAROUSEL_INDICATOR_SIZE = exports.DEFAULT_CAROUSEL_COLOR_INACTIVE = exports.DEFAULT_CAROUSEL_COLOR_ACTIVE = void 0;
exports.DEFAULT_CAROUSEL_COLOR_ACTIVE = '#a1a1a1';
exports.DEFAULT_CAROUSEL_COLOR_INACTIVE = '#d8d8d8';
exports.DEFAULT_CAROUSEL_INDICATOR_SIZE = 12;
const DEFAULT_CAROUSEL_INDICATOR_STYLE = {
    borderRadius: exports.DEFAULT_CAROUSEL_INDICATOR_SIZE / 2,
    height: exports.DEFAULT_CAROUSEL_INDICATOR_SIZE,
    margin: exports.DEFAULT_CAROUSEL_INDICATOR_SIZE / 3,
    width: exports.DEFAULT_CAROUSEL_INDICATOR_SIZE,
};
exports.DEFAULT_CAROUSEL_INDICATOR_ACTIVE_STYLE = {
    ...DEFAULT_CAROUSEL_INDICATOR_STYLE,
    backgroundColor: exports.DEFAULT_CAROUSEL_COLOR_ACTIVE,
};
exports.DEFAULT_CAROUSEL_INDICATOR_INACTIVE_STYLE = {
    ...DEFAULT_CAROUSEL_INDICATOR_STYLE,
    backgroundColor: exports.DEFAULT_CAROUSEL_COLOR_INACTIVE,
};
exports.VIEWABILITY_CONFIG = { itemVisiblePercentThreshold: 60 };
