// @aws-amplify/ui-react-native/lib/InAppMessaging/components/BannerMessage/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = exports.positionStyle = void 0;
const constants_1 = require("../../constants");
exports.positionStyle = {
    bottom: {
        justifyContent: 'flex-end',
    },
    middle: {
        justifyContent: 'center',
    },
    top: {
        justifyContent: 'flex-start',
    },
};
const getStyles = (imageDimensions, additionalStyle) => ({
    body: {
        fontSize: constants_1.FONT_SIZE_BASE,
        fontWeight: constants_1.FONT_WEIGHT_BASE,
        lineHeight: constants_1.LINE_HEIGHT_BASE,
    },
    buttonContainer: {
        backgroundColor: constants_1.COLOR_LIGHT_GREY,
        borderRadius: constants_1.BORDER_RADIUS_BASE,
        flex: 1,
        margin: constants_1.SPACING_MEDIUM,
        padding: constants_1.SPACING_LARGE,
    },
    buttonsContainer: {
        flexDirection: 'row',
        justifyContent: 'center',
        padding: constants_1.SPACING_SMALL,
    },
    buttonText: {
        fontSize: constants_1.FONT_SIZE_BASE,
        fontWeight: constants_1.FONT_WEIGHT_BASE,
        lineHeight: constants_1.LINE_HEIGHT_BASE,
        textAlign: 'center',
    },
    container: {
        backgroundColor: constants_1.COLOR_WHITE,
        elevation: constants_1.MESSAGE_ELEVATION,
        margin: constants_1.SPACING_EXTRA_LARGE,
        shadowColor: constants_1.COLOR_BLACK,
        shadowOffset: {
            width: constants_1.MESSAGE_SHADOW_WIDTH,
            height: constants_1.MESSAGE_SHADOW_HEIGHT,
        },
        shadowOpacity: constants_1.MESSAGE_SHADOW_OPACITY,
        shadowRadius: constants_1.MESSAGE_SHADOW_RADIUS,
    },
    contentContainer: {
        flexDirection: 'row',
        padding: constants_1.SPACING_LARGE,
    },
    header: {
        fontSize: constants_1.FONT_SIZE_LARGE,
        fontWeight: constants_1.FONT_WEIGHT_BOLD,
        lineHeight: constants_1.LINE_HEIGHT_LARGE,
    },
    iconButton: {
        alignSelf: 'flex-start',
        marginLeft: 'auto',
    },
    image: {
        ...imageDimensions,
    },
    imageContainer: {
        justifyContent: 'center',
    },
    textContainer: {
        flex: 1,
        paddingHorizontal: constants_1.SPACING_MEDIUM,
    },
    wrapper: {
        ...additionalStyle?.position,
        backgroundColor: 'transparent',
        flex: 1,
    },
});
exports.getStyles = getStyles;
